LiteCore/Storage/SQLiteKeyStore.cc

```diff
@@ -331,52 +331,103 @@ namespace litecore {
+            SQLite::Statement *stmt;
+            if (rec.sequence == 0_seq) {
+                // Insert only:
+                stmt = &compileCached(
+                        "INSERT OR IGNORE INTO kv_@ (version, body, extra, flags, sequence, key)"
+                        " VALUES (?, ?, ?, ?, ?, ?)");
+                opName = "insert";
+            } else {
+                // Replace only:
+                stmt = &compileCached(
+                        "UPDATE kv_@ SET version=?, body=?, extra=?, flags=?, sequence=?"
+                        " WHERE key=? AND sequence=? AND (flags >> 16) = ?");
+                stmt->bind(OldSequenceParam,    (long long)rec.sequence);
+                stmt->bind(OldSubsequenceParam, (long long)rec.subsequence);
+                opName = "update";
+            }
 
-        UsingStatement u(*stmt);
-        if (stmt->exec() == 0)
-            return 0_seq;               // condition wasn't met, i.e. conflict
+            sequence_t seq;
+            int64_t rawFlags = int(rec.flags);
+            if (updateSequence) {
+                seq = lastSequence() + 1;
+            } else {
+                Assert(rec.sequence > 0_seq);
+                seq = rec.sequence;
+                // If we don't update the sequence, update the subsequence so MVCC can work:
+                rawFlags |= (rec.subsequence + 1) << 16;
+            }
+
+            stmt->bindNoCopy(VersionParam, rec.version.buf, (int)rec.version.size);
+            stmt->bindNoCopy(BodyParam,    rec.body.buf, (int)rec.body.size);
+            stmt->bindNoCopy(ExtraParam,   rec.extra.buf, (int)rec.extra.size);
+            stmt->bind      (FlagsParam,   (long long)rawFlags);
+            stmt->bindNoCopy(KeyParam,     (const char*)rec.key.buf, (int)rec.key.size);
+            stmt->bind      (SequenceParam,(long long)seq);
+
+            if (db().willLog(LogLevel::Verbose) && name() != "default")
+                db()._logVerbose("KeyStore(%-s) %s %.*s", name().c_str(), opName, SPLAT(rec.key));
+
+            UsingStatement u(*stmt);
+            int status;
+            try {
+                status = stmt->exec();
+                // We are good, don't try again.
+                tryAgain = false;
+            } catch (const SQLite::Exception& exc) {
+                // We won't try again twice.
+                tryAgain = !tryAgain;
+                int sqliteErrorCode = exc.getErrorCode();
+                int sqliteExtErrorCode = exc.getExtendedErrorCode();
+                if (tryAgain && sqliteErrorCode == SQLITE_CONSTRAINT &&
+                    sqliteExtErrorCode == SQLITE_CONSTRAINT_UNIQUE) {
+                    // We only handle this error,
+                    // Log: Database | UNIQUE constraint failed: kv_default.sequence (19/2067)
+                    // Jot down the exception that makes us to try again.
+                    lastExcArgs = std::make_tuple(exc.what(), sqliteErrorCode, sqliteExtErrorCode);
```

Comment:
Could you just store the SQLite::Exception object itself, instead of having to break its parts into a tuple?